google/cloud/pubsublite/internal/resumable_async_streaming_read_write_rpc.h

```diff
@@ -160,48 +174,21 @@ class ResumableAsyncStreamingReadWriteRpcImpl
         case State::kShutdown:
           return make_ready_future(absl::optional<ResponseType>());
         case State::kRetrying:
+          assert(!read_reinit_done_.has_value());
           read_reinit_done_.emplace();
           return read_reinit_done_->get_future().then([](future<void>) {
             return make_ready_future(absl::optional<ResponseType>());
           });
         case State::kInitialized:
           read_future = stream_->Read();
+          assert(!in_progress_read_.has_value());
           in_progress_read_.emplace();
       }
     }
 
     return read_future.then(
         [this](future<absl::optional<ResponseType>> optional_response_future) {
-          absl::optional<promise<void>> in_progress_read;
-          future<void> read_reinit_done;
-          auto optional_response = optional_response_future.get();
-          bool shutdown;
-
-          {
-            std::lock_guard<std::mutex> g{mu_};
-            in_progress_read.swap(in_progress_read_);
-            shutdown = stream_state_ == State::kShutdown;
-            if (!optional_response.has_value() && !shutdown) {
-              read_reinit_done_.emplace();
-              read_reinit_done = read_reinit_done_->get_future();
-            }
-          }
-
-          in_progress_read->set_value();
-
-          if (shutdown) {
-            return make_ready_future(absl::optional<ResponseType>());
-          }
-
-          if (optional_response.has_value()) {
-            return make_ready_future(std::move(optional_response));
-          }
-
-          ReadWriteRetryFailedStream();
-
-          return read_reinit_done.then([](future<void>) {
-            return make_ready_future(absl::optional<ResponseType>());
-          });
+          return OnReadFutureFinish(std::move(optional_response_future));
```

Comment:
nit. make these functions accept the future value type and call .get() here and below